### ISSUE TITLE
"NEED" for HTTPS

### ISSUE THREAD

**Issue by aa-altieri** [yetanalytics/xapisec#2]
I appreciate the effort on this. However, I have to caution that we MUST allow for the OPTION of open HTTP statements to be sent. There may be times when security is not prioritized over mobility. For example, looking at IoT applications, many smaller systems (arduino-based systems, I'm looking at you!) simply cannot yet handle the full HTTPS stack implementation.

For those instances where being light-weight and agile are more important than hardened security (a weather station, for example, or how often a door is opened/closed), I think we must forbid ourselves from absolutely requiring HTTPS calls to be used exclusively.

I'm not at all disagreeing with your effort. Just the line that says that we should hold the xAPI spec to an HTTPS-only standard.

Just my two bits.

--anthony

**Comment by canweriotnow** [comment]
@aa-altieri This issue was actually discussed quite fervently and intensely on the Github issue for public comment on the White House CIO proposal. Other use cases for HTTP included public schools and libraries that use filtering (snooping) software that forbid HTTP traffic (for instance, I didn't realize many public schools can't access Google or YouTube b/c Google went HSTS across the board).

I think maybe the language should be clarified; xapisec as a proposal is for LRSs that want to consider themselves secure, hardened, what have you; kind of like SELinux. Sure, SELinux is a default on most distros now, but you can disable it (at your peril).

There are many solutions to the issues you describe; with beacons and arduinos and wearables, I've not seen them (typically) running a full HTTP stack on their own, but rather sending a byte stream to a "master" device (could be a phone, could be a RaspPi, could be a rack of Sun servers) that makes the actual connection to the LRS; in this case, if you have access to the full Java or C++ stdlib and OpenSSL, it's not really a problem. Or it could be a lightweight mini-LRS nearby that isn't "xapisec" and can do store-and-forward. 

So I'm not sure what the language change should be, as we do want to be emphatic about HTTPS and HSTS; the language was probably deliberately chosen to be overly emphatic on our part. 

But the issues you raise are also part of why we proposed xapisec as an independent set of best practices, outside the spec itself.

**Comment by blakeplock** [comment]
> xapisec as a proposal is for LRSs that want to consider themselves secure, hardened, what have you; kind of like SELinux. Sure, SELinux is a default on most distros now, but you can disable it (at your peril).

I like this. I think putting the decision into the equation will actually increase adoption of the protocol.

**Comment by aa-altieri** [comment]
I would, likewise, agree with that approach.  I guess my intention was to simply ensure that we're aware that not all LRS activity needs to be secured.  

However, Jason, you bring up an interesting point.  Consider the following (horrible) illustration:

Content ----HTTP---> LRS1 ---- HTTPS----> Main LRS  <-----HTTPS ---- LRS2 <---HTTPS --- Content

In this example, there are two bits of content.  One is sending HTTPS to an LRS (LRS2) which is forwarding HTTPS to the main LRS.  The other, is sending clear HTTP statements to an LRS (LRS1) which is forwarding as HTTPS.  Can this Main LRS, then, be said to be compliant?  

How far is the intentioned scope of this going to be?  I'm by no means suggestion either path.  But we do need to be a little careful about how well we define what we're certifying here.  If a supporting system is wide open, is the main LRS compliant?

Really, I know I'm mostly splitting hairs here a little.  But that's were these types of things get in trouble.

--anthony

**Comment by canweriotnow** [comment]
@aa-altieri Yes - I think in xAPI terms, LRS1 and LRS2 have to be considered as independent entities of "Main LRS" (Let's call it LRS0). LRS0 could be said to be compliant because it's responsible for its _own_ security and integrity, and has to remian agnostic about intermediaries to a point in order to still comply with the xAPI spec as it stands. 

That being said... I could envision a situation in which the "Authority" attribute of xAPI could be used to track the originating authority of a statement, which in common practice currently is the original receiving LRS. 

Could I envision a use case for a chain of trust in which Statements were rejected because of an untrusted originating authority? Sure, but I don't think that's what we're talking about here, but it might be desirable to some parties.

Could I envision a future xAPIsec model that followed the clearance model of secret/TS/TS-SCI levels of heightened restriction? Sure, but not yet :)

I think if anything, creative use of `Authority` or some other element to indicate "This LRS is xAPIsec but statement `some-UUID` passed through a non-xAPIsec LRS, just so you know" could open a path to make a world where non-TLS client/temp LRSs and secure LRSs can coexist and work together, in the model you describe. 

The effort isn't about eliminating possibilites for xAPI like that, it's about documenting a set of best practices for secure xAPI communication, storage, etc.

Again, this is early days, and these are the sort of issues we'll need to flesh out, and I hope you'll be able to participate once we set up some more calls etc. to discuss them. You've already given me some great ideas for adding sections to the document for conformance testing and trust relationships :smile:
